src/mol-canvas3d/passes/hi-z.ts

```diff
@@ -0,0 +1,559 @@
+    maxFrameLag: PD.Numeric(10, { min: 1, max: 30, step: 1 }, { description: 'Maximum number of frames to wait for Z-buffer data.' }),
+    minLevel: PD.Numeric(3, { min: 1, max: 10, step: 1 }),
+};
+export type HiZProps = PD.Values<typeof HiZParams>
+
+export class HiZPass {
+    private readonly viewport = Viewport();
+
+    private near = 0;
+    private far = 0;
+    private readonly view = Mat4();
+    private readonly projection = Mat4();
+
+    private nextNear = 0;
+    private nextFar = 0;
+    private readonly nextView = Mat4();
+    private readonly nextProjection = Mat4();
+
+    private readonly aabb = Vec4();
+    private readonly vp = Vec3();
+
+    private readonly levelData: LevelData = [];
+    private readonly fb: Framebuffer;
+    private readonly buf: WebGLBuffer;
+    private readonly tex: Texture;
+    private readonly renderable: HiZRenderable;
+    private readonly supported: boolean;
+
+    private sync: WebGLSync | null = null;
+    private buffer = new Float32Array(0);
+    private frameLag = 0;
+    private ready = false;
+
+    readonly props: HiZProps;
+
+    clear() {
+        if (!this.supported) return;
+
+        const { gl } = this.webgl;
+        if (!isWebGL2(gl)) return;
+
+        if (this.sync !== null) {
+            gl.deleteSync(this.sync);
+            this.sync = null;
+        }
+        this.frameLag = 0;
+        this.ready = false;
+
+        if (this.debug) {
+            this.debug.rect.style.display = 'none';
+            this.debug.container.style.display = 'none';
+        }
+    }
+
+    render(camera: Camera) {
+        if (!this.supported || !this.props.enabled) return;
+
+        const { gl, state } = this.webgl;
+        if (!isWebGL2(gl) || this.sync !== null) return;
+
+        this.nextNear = camera.near;
+        this.nextFar = camera.far;
+        Mat4.copy(this.nextView, camera.view);
+        Mat4.copy(this.nextProjection, camera.projection);
+
+        if (isTimingMode) this.webgl.timer.mark('hi-Z');
+
+        state.disable(gl.CULL_FACE);
+        state.disable(gl.BLEND);
+        state.disable(gl.DEPTH_TEST);
+        state.disable(gl.SCISSOR_TEST);
+        state.depthMask(false);
+        state.colorMask(true, true, true, true);
+        state.clearColor(0, 0, 0, 0);
+
+        //
+
+        const v = this.renderable.values;
+
+        for (let i = 0, il = this.levelData.length; i < il; ++i) {
+            const td = this.levelData[i];
+            td.framebuffer.bind();
+
+            if (i > 0) {
+                ValueCell.update(v.uInvSize, td.invSize);
+                ValueCell.update(v.uOffset, Vec2.set(v.uOffset.ref.value, 0, 0));
+                ValueCell.update(v.tPreviousLevel, this.levelData[i - 1].texture);
+            } else {
+                const s = Math.pow(2, Math.ceil(Math.log(Math.max(gl.drawingBufferWidth, gl.drawingBufferHeight)) / Math.log(2)) - 1);
```

Comment:
This is independent of the for loop, move it up?

Reply:
thanks, done